src/items/recipes.zig

```diff
@@ -0,0 +1,226 @@
+		newKeyPairs[0] = newKeys;
+		return newKeyPairs;
+	} else {
+		return error.NoMatch;
+	}
+}
+
+const ItemWithAmount = struct {
+	item: BaseItemIndex,
+	amount: u16,
+};
+
+const ItemKeyPair = struct {item: ItemWithAmount, keys: std.StringHashMap([]const u8)};
+
+fn findRecipeItemOptions(allocator: NeverFailingAllocator, itemStackPattern: ItemStackPattern, keys: *const std.StringHashMap([]const u8)) ![]ItemKeyPair {
+	const pattern = itemStackPattern.pattern;
+	const amount = itemStackPattern.amount;
+
+	if(pattern.len == 1 and pattern[0] == .literal) {
+		const item = BaseItemIndex.fromId(pattern[0].literal) orelse return error.ItemNotFound;
+		return allocator.dupe(ItemKeyPair, &.{.{
+			.item = .{
+				.item = item,
+				.amount = amount,
+			},
+			.keys = keys.clone() catch unreachable,
+		}});
+	}
+	var itemPairs: main.List(ItemKeyPair) = .initCapacity(allocator, 1);
+	defer itemPairs.deinit();
+	var iter = items.iterator();
+	while(iter.next()) |item| {
+		const newKeyMatches = matchWithKeys(allocator, item.id(), pattern, keys) catch continue;
+		for(newKeyMatches) |match| {
+			itemPairs.append(.{
+				.item = .{
+					.item = item.*,
+					.amount = amount,
+				},
+				.keys = match,
+			});
+		}
+	}
+	return itemPairs.toOwnedSlice();
+}
+
+fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []ZonElement) ![][]ItemWithAmount {
+	const arena = main.stackAllocator.createArena();
+	defer main.stackAllocator.destroyArena(arena);
+
+	var inputCombos: main.List([]ItemWithAmount) = .initCapacity(arena, 1);
+	inputCombos.append(arena.alloc(ItemWithAmount, recipe.len));
+	var keyList: main.List(std.StringHashMap([]const u8)) = .initCapacity(arena, 1);
+	keyList.append(.init(arena.allocator));
+	for(0.., recipe[0..]) |i, itemZon| {
+		const pattern = try parseItemZon(arena, itemZon);
+		var newKeyList: main.List(std.StringHashMap([]const u8)) = .init(arena);
+		var newInputCombos: main.List([]ItemWithAmount) = .init(arena);
+
+		for(keyList.items, inputCombos.items) |*keys, inputs| {
+			const parsedItems = try findRecipeItemOptions(arena, pattern, keys);
+			for(parsedItems) |item| {
+				const newInputs = arena.dupe(ItemWithAmount, inputs);
+				newInputs[i] = item.item;
+				newInputCombos.append(newInputs);
+				newKeyList.append(item.keys);
+			}
+		}
+		keyList = newKeyList;
+		inputCombos = newInputCombos;
+	}
+	const newInputCombos = allocator.alloc([]ItemWithAmount, inputCombos.items.len);
+	for(inputCombos.items, 0..) |inputCombo, i| {
+		newInputCombos[i] = allocator.dupe(ItemWithAmount, inputCombo);
+	}
+	return newInputCombos;
+}
+
+pub fn addRecipe(itemCombo: []const ItemWithAmount, list: *main.List(Recipe)) void {
+	const inputs = itemCombo[0 .. itemCombo.len - 1];
+	const output = itemCombo[itemCombo.len - 1];
+	const recipe = Recipe{
+		.sourceItems = main.globalAllocator.alloc(BaseItemIndex, inputs.len),
+		.sourceAmounts = main.globalAllocator.alloc(u16, inputs.len),
```

Comment:
Since this is freed in a completely different file, I think it would be better to pass the allocator as a function argument.
This reduces the amount of work when migrating to a different allocator and reduces the risk for accidental leaks when doing so.